File: util/astar.js
const helpers = require("./helpers");
const { GameState, OpenCell } = require("./game_state");

const VERY_HIGH_COST = 1000;

function getIndex(coord, gameState) {
	return helpers.cellIndex(coord.x, coord.y, gameState.width);
}

// Just calculate the distance if there were no barriers.
function heuristicCostEstimate(start, goal) {
	return Math.abs(start.x - goal.x) + Math.abs(start.y - goal.y);
}

function lowestFScoreInSet(set, fScore) {
	let lowest = null;
	let lowestIndex = null;
	set.forEach((index) => {
		const score = fScore[index];
		if (score !== undefined && score !== null) {
			if (lowest === null || score < lowest) {
				lowest = score;
				lowestIndex = index;
			}
		}
	});
	return lowestIndex;
}

function reconstructPath(cameFrom, currentIndex, width) {
	const moves = [];
	while (currentIndex in cameFrom) {
		const nextIndex = cameFrom[currentIndex];
		const direction = helpers.directionBetweenNodes(nextIndex, currentIndex, width);
		moves.unshift(direction);
		currentIndex = nextIndex;
	}
	return moves;
}

function indexIsSafe(index, turn, gameState) {
	const coord = helpers.deconstructCellIndex(index, gameState.width);
	if (helpers.outOfBounds(coord, gameState)) {
		return false;
	}
	const turnsUntilCouldBeVacant = gameState.map.turnsUntilVacant(coord);
	return turnsUntilCouldBeVacant < turn;
	// const myTail = gameState.mySnake.tail();
	// const isMyTail = myTail.x === x && myTail.y === y;
	// return isMyTail || gameState.checkCell(x, y).constructor === OpenCell;
}

function isAdjacent(aIndex, bIndex, gameState) {
	const a = helpers.deconstructCellIndex(aIndex, gameState.width);
	const b = helpers.deconstructCellIndex(bIndex, gameState.width);

	const horizontallyAdjacent = Math.abs(a.x - b.x) === 1 && a.y === b.y;
	const verticallyAdjacent = Math.abs(a.y - b.y) === 1 && a.x === b.x;

	return horizontallyAdjacent || verticallyAdjacent;
}

// This is for the special case where the AI will turn back on itself in the second move
// of the game when it is only two cells long. That's an illegal move even though the
// tail will no longer be in that position after the move.
function is180(index, neighborIndex, gameState) {
	const head = gameState.mySnake.parts[0];
	const neck = gameState.mySnake.parts[1];

	if (!neck) {
		// just in case length of 1 is allowed
		return false;
	}

	const headIndex = helpers.cellIndex(head.x, head.y, gameState.width);
	const neckIndex = helpers.cellIndex(neck.x, neck.y, gameState.width);
	return index === headIndex && neighborIndex == neckIndex;
}

function isCloseToHead(index, snake, gameState) {
	const headIndex = helpers.cellIndex(snake.head().x, snake.head().y, gameState.width);
	return isAdjacent(index, headIndex, gameState);
}

function isCloseToEqualOrBiggerSnakeHead(index, gameState) {
	for (const otherSnake of gameState.enemies) {
		const otherSnakeIsTooBigToEat = gameState.mySnake.length() <= otherSnake.length();
		if (otherSnakeIsTooBigToEat && isCloseToHead(index, otherSnake, gameState)) {
			return true;
		}
	}
	return false;
}

function getNeighbors(index, turn, gameState) {
	const coord = helpers.deconstructCellIndex(index, gameState.width);
	const result = [];

	function addIfSafe(i) {
		if (indexIsSafe(i, turn, gameState) && !is180(index, i, gameState) && isAdjacent(index, i, gameState)) {
			result.push(i);
		}
	}

	addIfSafe(index + 1); // right
	addIfSafe(index - 1); // left
	addIfSafe(index + gameState.width); // down
	addIfSafe(index - gameState.width); // up

	return result;
}

// Partially stolen from the pseudo-code in the wikipedia page for A*:
// https://en.wikipedia.org/wiki/A*_search_algorithm#Pseudocode
function shortestPath(start, goal, gameState) {

	const startIndex = getIndex(start, gameState);
	const goalIndex = getIndex(goal, gameState);

	// Allow some special functionality on first move to avoid being eaten.
	let isFirstMove = true;

	// Nodes/cells that have already been evaluated.
	const closedSet = new Set();

	// Nodes/cells that are not yet evaluated but have been discovered. Only the start
	// node is discovered so far.
	const openSet = new Set([ startIndex ]);

	// For each node store the node it can most efficiently be reached from.
	const cameFrom = {};

	// For each node the cost of getting from the start to that node.
	const gScore = {};

	// Like gScore but just straight up turn count to get to each cell. gScore could
	// be exagerrated to discourage certain risky paths.
	const turns = {};

	// Already at start so cost of going there is zero.
	gScore[startIndex] = 0;
	
	// Start at turn 1 because it makes more sense in my head.
	turns[startIndex] = 1;

	// For each node the cost of getting from start to goal passing through that node.
	// Depends on heuristics. Not guaranteed to be precise.
	const fScore = {};

	// fScore for start node is 100% heuristic.
	fScore[startIndex] = heuristicCostEstimate(start, goal);

	while (openSet.size > 0) {
		const currentIndex = lowestFScoreInSet(openSet, fScore);

		if (currentIndex === goalIndex) {
			return reconstructPath(cameFrom, currentIndex, gameState.width);
		}

		openSet.delete(currentIndex);
		closedSet.add(currentIndex);

		const neighbors = getNeighbors(currentIndex, turns[currentIndex], gameState);
		for (const neighborIndex of neighbors) {
			// Don't revisit the same node we've already been to.
			if (closedSet.has(neighborIndex)) {
				continue;
			}

			// Check if we have discovered a new node.
			if (!openSet.has(neighborIndex)) {
				openSet.add(neighborIndex);
			}

			// On immediate move avoid getting eaten.
			let dontGetEatenModifier = 0;
			if (isFirstMove) {
				const couldGetEaten = isCloseToEqualOrBiggerSnakeHead(neighborIndex, gameState);
				if (couldGetEaten) {
					dontGetEatenModifier = VERY_HIGH_COST;
				}
			}

			// Distance between neighbors is always 1, hence the +1. Also if there is a chance that
			// you could get eaten in this sport then consider it much higher cost.
			const tentativeGScore = gScore[currentIndex] + 1 + dontGetEatenModifier;

			// If there's already a better gScore for this node then we're not on the
			// ideal path so bail.
			if (tentativeGScore >= gScore[neighborIndex]) {
				continue;
			}

			// For now this is the best path to this node so record it.
			cameFrom[neighborIndex] = currentIndex;
			gScore[neighborIndex] = tentativeGScore;
			turns[neighborIndex] = turns[currentIndex] + 1;
			fScore[neighborIndex] = gScore[neighborIndex] + heuristicCostEstimate(
				helpers.deconstructCellIndex(neighborIndex, gameState.width), goal);
		}

		isFirstMove = false;
	}

	// Finished searching and could not find a path :(
	return null;
}

module.exports = {
	shortestPath
};

File: napi/astar.cpp
#include "astar.hpp"

#include <algorithm>
#include <unordered_set>
#include <unordered_map>

#define VERY_HIGH_COST 1000

uint32_t heuristicCostEstimate(Point start, Point goal)
{
    return absDiff(start.x, goal.x) + absDiff(start.y, goal.y);
}

uint32_t lowestFScoreInSet(
    std::unordered_set<uint32_t> &set, std::unordered_map<uint32_t, uint32_t> &fScore)
{
    uint32_t lowest = 0;
    uint32_t lowestIndex = 0;
    bool first = true;

    for (auto index : set)
    {
        auto iter = fScore.find(index);
        if (iter != fScore.end())
        {
            uint32_t score = iter->second;
            if (first || score < lowest)
            {
                lowest = score;
                lowestIndex = index;
                first = false;
            }
        }
    }

    return lowestIndex;
}

std::vector<Direction> reconstructPath(
    std::unordered_map<uint32_t, uint32_t> &cameFrom, uint32_t currentIndex, uint32_t width)
{
    std::vector<Direction> moves;
    auto iter = cameFrom.find(currentIndex);
    while (iter != cameFrom.end())
    {
        uint32_t nextIndex = iter->second;
        Direction direction = directionBetweenNodes(nextIndex, currentIndex, width);
        moves.insert(moves.begin(), direction);
        currentIndex = nextIndex;
        iter = cameFrom.find(currentIndex);
    }

    return moves;
}

bool indexIsSafe(uint32_t index, uint32_t turn, GameState &state)
{
    Point coord = deconstructCellIndex(index, state);
    if (outOfBounds(coord, state))
    {
        return false;
    }
    uint32_t turnsUntilCouldBeVacant = state.map().turnsUntilVacant(coord);
    return turnsUntilCouldBeVacant < turn;
}

bool isAdjacent(uint32_t aIndex, uint32_t bIndex, GameState &state)
{
    Point a = deconstructCellIndex(aIndex, state);
    Point b = deconstructCellIndex(bIndex, state);

    bool horizontallyAdjacent = absDiff(a.x, b.x) == 1 && a.y == b.y;
    bool verticallyAdjacent = absDiff(a.y, b.y) == 1 && a.x == b.x;

    return horizontallyAdjacent || verticallyAdjacent;
}

bool is180(uint32_t index, uint32_t neighborIndex, GameState &state)
{
    // If length of snake is 1 then nothing to worry about (... I think... might wanna test that)
    if (state.mySnake()->length() <= 1)
    {
        return false;
    }

    Point head = state.mySnake()->parts.at(0);
    Point neck = state.mySnake()->parts.at(1);

    uint32_t headIndex = cellIndex(head, state);
    uint32_t neckIndex = cellIndex(neck, state);

    return index == headIndex && neighborIndex == neckIndex;
}

bool isCloseToHead(uint32_t index, Snake *snake, GameState &state)
{
    uint32_t headIndex = cellIndex(snake->head(), state);
    return isAdjacent(index, headIndex, state);
}

bool isCloseToEqualOrBiggerSnakeHead(uint32_t index, GameState &state)
{
    auto enemies = state.enemies();
    for (Snake *otherSnake : enemies)
    {
        bool otherSnakeIsTooBigToEat = state.mySnake()->length() <= otherSnake->length();
        if (otherSnakeIsTooBigToEat && isCloseToHead(index, otherSnake, state))
        {
            return true;
        }
    }
    return false;
}

bool isOkNeighbor(uint32_t index, uint32_t other, uint32_t turn, GameState &state)
{
    bool result = indexIsSafe(other, turn, state)
        && !is180(index, other, state)
        && isAdjacent(index, other, state);

    return result;
}

std::vector<uint32_t> getNeighbors(uint32_t index, uint32_t turn, GameState &state)
{
    std::vector<uint32_t> result;

    uint32_t right = index + 1;
    uint32_t left = index - 1;
    uint32_t down = index + state.width();
    uint32_t up = index - state.width();

    if (isOkNeighbor(index, up, turn, state)) result.push_back(up);
    if (isOkNeighbor(index, down, turn, state)) result.push_back(down);
    if (isOkNeighbor(index, left, turn, state)) result.push_back(left);
    if (isOkNeighbor(index, right, turn, state)) result.push_back(right);

    return result;
}

uint32_t getGScore(std::unordered_map<uint32_t, uint32_t> &gScore, uint32_t index)
{
    auto iter = gScore.find(index);
    if (iter == gScore.end())
    {
        return VERY_HIGH_COST;
    }
    else
    {
        return iter->second;
    }
}

std::vector<Direction> shortestPath(Point start, Point goal, GameState &state)
{
    std::cout << "***2.1.1\n";

    uint32_t startIndex = cellIndex(start, state);
    uint32_t goalIndex = cellIndex(goal, state);

    bool isFirstMove = false;
    std::unordered_set<uint32_t> closedSet;
    std::unordered_set<uint32_t> openSet;
    std::unordered_map<uint32_t, uint32_t> cameFrom;
    std::unordered_map<uint32_t, uint32_t> gScore;
    std::unordered_map<uint32_t, uint32_t> fScore;
    std::unordered_map<uint32_t, uint32_t> turns;

    std::cout << "***2.1.2\n";

    openSet.insert(startIndex);
    std::cout << "***2.1.2.1\n";
    gScore.emplace(startIndex, 0);
    std::cout << "***2.1.2.2\n";
    auto estimate = heuristicCostEstimate(start, goal);
    fScore.emplace(startIndex, estimate);
    std::cout << "***2.1.2.3\n";
    turns.emplace(startIndex, 1);

    std::cout << "***2.1.3\n";

    while (!openSet.empty())
    {
        std::cout << "***2.1.4\n";
        uint32_t currentIndex = lowestFScoreInSet(openSet, fScore);
        std::cout << "***2.1.5\n";
        if (currentIndex == goalIndex)
        {
            std::vector<Direction> path = reconstructPath(cameFrom, currentIndex, state.width());
            return path;
        }

        openSet.erase(currentIndex);
        closedSet.insert(currentIndex);

        std::vector<uint32_t> neighbors = getNeighbors(currentIndex, turns[currentIndex], state);
        for (uint32_t neighborIndex : neighbors)
        {
            std::cout << "***2.1.6\n";
            if (closedSet.find(neighborIndex) != closedSet.end())
            {
                continue;
            }

            if (openSet.find(neighborIndex) == openSet.end())
            {
                openSet.insert(neighborIndex);
            }

            uint32_t dontGetEatenModifier = 0;
            if (isFirstMove)
            {
                bool couldGetEaten = isCloseToEqualOrBiggerSnakeHead(neighborIndex, state);
                if (couldGetEaten)
                {
                    dontGetEatenModifier = VERY_HIGH_COST;
                }
            }

            uint32_t tentativeGScore = getGScore(gScore, currentIndex) + 1 + dontGetEatenModifier;

            if (tentativeGScore >= getGScore(gScore, neighborIndex))
            {
                continue;
            }

            cameFrom[neighborIndex] = currentIndex;
            gScore[neighborIndex] = tentativeGScore;
            turns[neighborIndex] = turns[currentIndex] + 1;
            fScore[neighborIndex] = gScore[neighborIndex] + heuristicCostEstimate(
                deconstructCellIndex(neighborIndex, state), goal);
        }

        isFirstMove = false;
    }

    // Found no path
    return std::vector<Direction>();
}